### DISCUSSION TITLE
Home-row Shift False Negatives at Start of Typing -- "fi" instead of "I"

### DISCUSSION THREAD

**Discussion by jeffsf** [urob/zmk-config#71] [Q&A] [answered]
In general, the home-row mods including Shift are working well for me at my modest typing speed (30-40 WPM and choppy).
However, I seem to get a high number of false negatives when starting after a pause, such as a sentence that begins with "I" resulting in the HRM key "f" and then the unmodified "i" being sent. It almost always seems to happen after a pause and with cross-hand sequences.

Reducing `require-prior-idle-ms` as suggested seems counterintuitive as there is generally a long idle before starting the two-key sequence when it does not provide the shifted key. Based on "you want to set it to at least 10500/x where x is your (relaxed) WPM for English prose." it is already shorter than the ~260 ms that would suggest. 

I'm using the `studio/keymap-subsystem` branch from `petejohanson/zmk` at a commit from June 20th, that seems to have been subsequently rewritten. 

Hopefully properly configured from your original sources:

```
/* left-hand HRMs */
ZMK_HOLD_TAP(hml,
    flavor = "balanced";
    tapping-term-ms = <280>;
    quick-tap-ms = <175>;                // repeat on tap-into-hold
    require-prior-idle-ms = <150>;
    bindings = <&kp>, <&kp>;
    hold-trigger-key-positions = <KEYS_R THUMBS>;
    hold-trigger-on-release;             // delay positional check until key-release
)

/* right-hand HRMs */
ZMK_HOLD_TAP(hmr,
    flavor = "balanced";
    tapping-term-ms = <280>;
    quick-tap-ms = <175>;                // repeat on tap-into-hold
    require-prior-idle-ms = <150>;
    bindings = <&kp>, <&kp>;
    hold-trigger-key-positions = <KEYS_L THUMBS>;
    hold-trigger-on-release;             // delay positional check until key-release
)
```

**Comment by urob** [comment, marked as answer]
If this occurs after pausing and for cross-hand mods, the `require-prior-idle-ms` setting doesn't matter here. Unless there is some bug in the firmware, my only guess would be that you aren't properly nesting the alpha key. In order for it to be shifted, one of two conditions must be satisfied:

1. Either the shift-HRM is held past `tapping-term` *before* pressing `i`. The underlying design principle behind timerless home-row mods makes this unlikely.
2. Or, you must properly nest the alpha in order to bypass the high tapping-term. In your case that is, press `f` , press and release `i`, release `f`. This is the intended way to trigger mods given the timerless design. 

When typing fast, it can be hard to meet either condition for the purpose of shifting alphas. This is the reason why many prefer to have a separate sticky-shift on one of the thumb keys for the purpose of shifting. 

If this is not an option, you could try to either lower the `tapping-term` or to switch to a `hold-preferred` flavor, but both are somewhat at odds with the timerless design and may have other unintended side effects.

**Comment by urob** [comment]
Just to be sure: your snippet above doesn't include a definition for `KEYS_R`, `KEYS_L` and  `THUMBS`. Unless you have them defined elsewhere, even your cross-hand mods would need to wait past the `tapping-term` (even when properly nested).

**Reply by jeffsf** [reply]
> your snippet above doesn't include a definition for `KEYS_R`, `KEYS_L` and `THUMBS`

Thanks, defined elsewhere.

My guess, based on your (1) and (2), is that the `f` is being released prior to the `i` being released. Confirmed

```
KEY-DOWN - QMK: KC_F    Event key: f           Code: KeyF          KeyCode: 70
KEY-DOWN - QMK: KC_I    Event key: i           Code: KeyI          KeyCode: 73
KEY-UP   - QMK: KC_F    Event key: f           Code: KeyF          KeyCode: 70 in 7.700ms
KEY-UP   - QMK: KC_I    Event key: i           Code: KeyI          KeyCode: 73 in 51.500ms
```
